### DISCUSSION TITLE
[Q] Intended use of Optional<T> and serialization problems

### DISCUSSION THREAD

**Discussion by ronnyek** [ChilliCream/hotchocolate#5028] [General]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

Having a Optional<T> parameter on a model I share between input and objecttypes causes an error in hotchocolate runtime
`Boolean cannot serialize the given value.`

Should `Optional<T>` ONLY be used in input type objects alone? Should they even be something I'd be including in any of my classes?

### Steps to reproduce

1. Modify original quick start sample (book/author)
``` cs 
// Book.cs
public class Book
    {
        public string Title { get; set; }
        public Optional<bool?> IsFavorite { get; set; }
        public Author Author { get; set; }
    }
```

``` cs
//Query.cs
public class Query
    {
        public Book GetBook() => new Book
        {
            Title = "C# In Depth",
            Author = new Author
            {
                Name = "John Skeet",
            }
        };
    }
```
2. Query for book information
``` js
    {
  book{
    title
    isFavorite
    author{
      name
    }
  }
}
```
3. Observe errors in response
``` js
{
  "errors": [
    {
      "message": "Boolean cannot serialize the given value.",
      "locations": [
        {
          "line": 4,
          "column": 5
        }
      ],
      "path": [
        "book",
        "isFavorite"
      ],
      "extensions": {
        "actualValue": "unspecified",
        "actualType": "HotChocolate.Optional`1[[System.Nullable`1[[System.Boolean, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]]",
        "code": "EXEC_INVALID_LEAF_VALUE"
      }
    }
  ],
```

### Additional Context?

I feel like mention in documentation on `Optional<T>` is minimal, and I'm unclear HOW they should be used. Should this be purely somethign that exists in input types themselves and no where else? If that's the case, why does hotchocolate require you to provide a value for an `Optional<bool>` or provide a default? Seems like Optional in the name means providing it should be optional, and that within the query, the optional value would just not be specified or whatever.

### Product

Hot Chocolate

### Version

12.7.0

**Comment by ronnyek** [comment]
Is optional even really supported anymore? Seems like documentation referenced it once that I could find, and its brief mentioned showed it being used in method parameters, or as properties on input objects...

If its supported, I'd love to see a real world application... or if someone could give me an idea of how people build mutation input types with fields that can be optionally provided, and how the wire that up... I'd be happy to contribute documentation

**Comment by KreativJos** [comment]
Optional is documented here: https://chillicream.com/docs/hotchocolate/v12/defining-a-schema/input-object-types/#optional-properties
Why would one want to use it in an output type? The only case I can think of is when using the same object for input & output.
